### PR TITLE
reader_concurrency_semaphore: execution_loop(): trigger admission check when _ready_list is empty

### PR DESCRIPTION
**Pull request by denesb**
The execution loop consumes permits from the _ready_list and executes them. The _ready_list usually contains a single permit. When the _ready_list is not empty, new permits are queued until it becomes empty. The execution loops relies on admission checks triggered by the read releasing resouces, to bring in any queued read into the _ready_list, while it is executing the current read. But in some cases the current read might not free any resorces and thus fail to trigger an admission check and the currently queued permits will sit in the queue until another source triggers an admission check.
I don't yet know how this situation can occur, if at all, but it is reproducible with a simple unit test, so it is best to cover this corner-case in the off-chance it happens in the wild. Add an explicit admission check to the execution loop, after the _ready_list is exhausted, to make sure any waiters that can be admitted with an empty _ready_list are admitted immediately and execution continues.

Fixes: #13540

### CODE REVIEW

File: test/boost/reader_concurrency_semaphore_test.cc [hunk 1]
```diff
@@ -1896,3 +1896,43 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_necessary_evicting) {
...
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 2);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted, 1);
+
+    // After func runs, the _ready_list becomes empty and the waiting permit should be admitted
+    func_fut.get();
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 0);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted, 2);
+
+    permit2_fut.get();
+}
```
[the remark this answers is not in the document]
Comment (denesb) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: User code should not be calling `maybe_admit_waiters()`, in fact it is a private method.

File: reader_concurrency_semaphore.cc [hunk 2]
```diff
@@ -907,6 +907,8 @@ future<> reader_concurrency_semaphore::execution_loop() noexcept {
                 co_await coroutine::maybe_yield();
             }
         }
+
+        maybe_admit_waiters();
```
[the remark this answers is not in the document]
Comment (denesb) [reply reader_concurrency_semaphore.cc:911]: Not sure what you mean. I added this to after the loop exhausts the ready list, just before it starts waiting on the conditional variable. This check ensures that we only go to sleep if no permit can be admitted.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (denesb) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: The only way I could reproduce the stalling is if the read doesn't release any release (doesn't call `signal()`). If it does, the stall doesn't reproduce, as `signal()` calls `maybe_admit_waiters()`.

File: reader_concurrency_semaphore.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply reader_concurrency_semaphore.cc:911]: Suppose the condition that allows us to admit a read becomes true only after this new check. So we have this sequence:

 - a read gets queued
 - this check tries to admit it, but fails
 - the condition that allows admitting the read becomes true
 - we're still asleep

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: But how can that happen? Every read releases a resource (the permit itself).

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: Do we call maybe_admit_waiters there? We should, since cached queries are eligible for eviction if a new query comes along.

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: I think we do. on_permit_inactive() eventually calls maybe_admit_waiters().

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]:
```c++
void reader_concurrency_semaphore::on_permit_destroyed(reader_permit::impl& permit) noexcept {
    permit.unlink();
    _permit_gate.leave();
    --_stats.current_permits;
    if (_blessed_permit == &permit) {
        _blessed_permit = nullptr;
        maybe_admit_waiters();
    }
}
```

shouldn't we admit even if we're not the blessed permit? Under non-oom conditions, there isn't a blessed permit, is there?

**Comment by denesb** [comment]
Ping.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: So isn't this the bug? Shouldn't register_inactive_read say to itself, here I am idle, holding on to resources that could be used for an active read, shouldn't I check the queue?

[the remark this answers is not in the document]
Comment (denesb) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: If a read is waiting on `_ready_list` to become empty, then it was already cleared w.r.t. resource availability. Dropping the inactive read will not help that waiter get admitted. In fact we used to unconditionally drop freshly registered inactive reads when there were waiters and I recently changed it because it wasted reads and lead to timeouts.

File: reader_concurrency_semaphore.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (denesb) [reply reader_concurrency_semaphore.cc:911]: `_ready_list` is always non-empty when the execution loop runs. If it is empty, the loop waits on the condition variable. The expectation is that by the time the current read finishes running, there will be another read pushed onto the `_ready_list`, via calls to `maybe_admit_waiters()`, called from the current read releasing resources. If there is no new read on the list, the loop waits on the condition variable. The current patch adds a forced poll for ready reads just before the loop goes to sleep, in case the previous read didn't trigger any (although that is very unlikely, but not impossible).

[the remark this answers is not in the document]
Comment (avikivity) [reply reader_concurrency_semaphore.cc:911]: How is it possible, barring a bug?

[the remark this answers is not in the document]
Comment (avikivity) [reply reader_concurrency_semaphore.cc:911]: I guess there's a bug in how the condition variable is used. Waits should always be in a look with a check just before, or using the overload of when() that accepts a predicate.

[the remark this answers is not in the document]
Comment (avikivity) [reply reader_concurrency_semaphore.cc:911]: But I think the code as written (before this patch) is equivalent. If there is a spurious wakeup, the while () loop will ignore it.

I still don't understand this patch.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: > If a read is waiting on `_ready_list` to become empty, then it was already cleared w.r.t. resource availability. 

Yes. I guess that's *at most* one new read, and any number of resumed reads.

Sounds like we have a small but benign race:
 - move a read from _waiting to _ready
 - wake the loop
 - resume some reads, moving them to _ready and waking the loop
 - the loop awakes, but it would have been happy with just the resumed reads, we added extra load for no reason

> Dropping the inactive read will not help that waiter get admitted. In fact we used to unconditionally drop freshly registered inactive reads when there were waiters and I recently changed it because it wasted reads and lead to timeouts.

Yes. But going back to this patch. It looks like a random place to call maybe_admit_waiters(). The places to call maybe_admit_waiters() are the ones that change the inputs to its computation. Is this place changing the input to the computation?

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: And indeed, 

```c++
    if (!_ready_list.empty()) {
        return {can_admit::no, reason::ready_list};
    }

```

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: So I now agree with the patch, but I think _ready is too complicated, and wrong, because it creates a gap between the time we calculate whether we should start a read, and actually starting it. During that time things can change. I don't think it's critical (certainly not compared to not launching a read, as does this bug), but it makes understanding mode difficult.

[the remark this answers is not in the document]
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1938]: I'll queue it, though I don't like the unit test.

File: reader_concurrency_semaphore.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (denesb) [reply reader_concurrency_semaphore.cc:911]: > How is it possible, barring a bug?

I don't know how this could happen in the real world. This patch is quite speculative to be honest. I could write a unit test reproducer but it is quite artificial. After all, what real read could complete a page and not release any resources? The only case I could think of is a very last page, where the page turns out to be empty and nothing was read. But in this case the reader (and the permit with it) would be destroyed, which releases the base resources of the permit.
